In [14]:
import pandas as pd
import numpy as np #storing as array
from sklearn import tree
from sklearn import preprocessing 
# used for finding missing values, text into numrical

In [15]:
df = pd.read_excel(r'C:\Users\mansi\Documents\LetsUpgrade\Day 24\Decision_Tree_Bank_Personal_Loan\Bank_Personal_Loan_Modelling (1).xlsx', sheet_name='Data')
df.head()

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   

   Personal Loan  Securities Account  CD Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

In [16]:
from sklearn.ensemble import RandomForestClassifier
model = RandomForestClassifier(n_estimators = 2000, max_features = 2, oob_score = True)

In [17]:
bank_train_feats =[ 'Age', 'Experience', 'Income', 'Family', 'CCAvg',
       'Education', 'Mortgage', 'Securities Account',
       'CD Account', 'Online', 'CreditCard']

In [31]:
df.isna().sum().sum()

0

In [18]:
model.fit(X = df[bank_train_feats], y = df['Personal Loan'])

RandomForestClassifier(max_features=2, n_estimators=2000, oob_score=True)

In [19]:
model.oob_score_

0.9876

In [20]:
for train_feat, imp in zip(bank_train_feats, model.feature_importances_):
    print(train_feat, '-->', imp)

Age --> 0.04454875941529665
Experience --> 0.04443198573090121
Income --> 0.342549109037465
Family --> 0.09705625890557831
CCAvg --> 0.18464408587357015
Education --> 0.16597648840987297
Mortgage --> 0.04335193812080511
Securities Account --> 0.005499021105199829
CD Account --> 0.0538410314451551
Online --> 0.00835742828381336
CreditCard --> 0.009743893672342222


##  CCAvg, Education, Income seem to be important features are usefull features

In [25]:
bank_test_feats = ['CCAvg', 'Education', 'Income']

In [26]:
model = tree.DecisionTreeClassifier(max_depth = 6)

In [27]:
model.fit(X = df[bank_test_feats], y = df['Personal Loan'])

DecisionTreeClassifier(max_depth=6)

In [28]:
model.score(X = df[bank_test_feats], y = df['Personal Loan'])

0.9738

In [30]:
with open("decision_tree.dot",'w')as f:
    f=tree.export_graphviz(model,feature_names=["CCAvg","Education","Income"],out_file=f);

## conclusion:
1 Person having CCAvg below ,25 and education below 1.5 have got loan.
    
6 people having  CCAvg below 1.65 and education below 1.5 have got loan.

2 people having income lesser than 108.5 and CCAvg between 1.5 and 1.65 have got loan.

3 people having CCAvg  lesser than 3.45 and income 82.5 have got loan.

12 people having CCAVG below 4.05 and education below 1.5 and income between 82.5 and 113.5 have got loan.

5 people having education below 2.5 and  income 92.5 and education below 1.5 have got loan.

18 people having income below 159.5 and education below 1.5 and CCAvg 4.05 have got loan.

29 people having income 179.5 and CCAvg between 6.55 and 4.05 have got loan.

1 person having CCAvg between 6.85 and 6.95 and income 176.5 have got loan.

7 person having CCAvg below 1.1 and income below 116.5 have got loan.

3 person having CCAvg below 3.35 and income below 116.5  have got loan.

